# Predicting Credit Scores
### By: Ayuj Verma, Gauri Pala, Ayush Bhalavat, Arnav Chopra

# Introduction
For our project, we are analyzing the Credit Score dataset from Kaggle, found [here](https://www.kaggle.com/datasets/parisrohan/credit-score-classification/data).

We are using features that relate to bank details and credit-related information for a customer to classify their credit score into a credit score bracket, either Good, Bad, or Standard. In this project, we aim to tackle the machine learning problem of credit score classfiication. This problem is important for various financial instiutions, like banks, and other lending practices that tend to use credit scores. A credit score essentially represents an individual's creditworthiness, indicating the likelihood of them repaying borrowed money on time based on information from your credit reports.

This problem is relevant as it impacts various aspects of financial-decision making. Many financial institutions heavily rely on credit scores to assess the risk associated with lending money to different individuals. In general, a higher credit score typically suggests a lower risk of default (considered Good), whereas a low credit score illustrates a higher risk. Therefore, accurate classification of credit scores into bins of Good, Bad, or Standard is crucial for lenders to make informed decisions about loan approvals and interest rates. Places like banks need to know if someone is likley to pay them back before they give out loans or credit cards, so they can give out loans more safely and avoid losing money.

Our model that classifies a predict's if a person has a good, bad, or standard credit score can be utilized for several purposes. For loan approval, lenders can use the model to automate the process of loan approvals as they can use the prediction as an indicator of whether this person should receive a loan or not, which can streamline the operations and reduces manual workload. Similarly, other financial institutions can better assess the risk associated with lending to individuals, leading to more accurate loan prices and mitigate potential money losses. Furthermore, the model can assist in determining appropriate interest rates for loans based on the creditworthiness of applicants. Overall, with this machine learning model, the process of making decisions based on someone's credit related information can be automated. 

Predicting credit score does matter because knowing someone's credit score helps banks understand how risky it is to lend them money, and if they can predict it accurately, they can offer fair interest rates and help more peolpe access loans and credit cards.

Our dataset includes a vast amount of a person's basic bank details and credit-related information. The dataset is setup in a way where multiple records are attached to the same customer. Essentially, each customer has 8 records associated with them: one record is associated with the customer's credit information for a particular month and every customer has records for January - August (8 months). Because the dataset is structured this way, it highlights how a person's credit score can change from, for instance, from Good to Standard, in a month if there is a change to the customer's information (open up more bank accounts one month, lose their job which means no income, forget to pay a statement).

The features included in the dataset are listed below:

ID: Represents a unique identification of an entry                              
Customer_ID: Represents a unique identification of a person                   
Month: Represents the month of the year                           
Name: Represents the name of a person                     
Age: Represents the age of the person                             
SSN: Represents the social security number of a person                             
Occupation: Represents the occupation of the person                       
Annual_Income: Represents the annual income of the person                   
Monthly_Inhand_Salary: Represents the monthly base salary of a person       
Num_Bank_Accounts: Represents the number of bank accounts a person holds                
Num_Credit_Card: Represents the number of other credit cards held by a person                 
Interest_Rate: Represents the interest rate on credit card                    
Num_of_Loan: Represents the number of loans taken from the bank                     
Type_of_Loan: Represents the types of loan taken by a person                
Delay_from_due_date: Represents the average number of days delayed from the payment date             
Num_of_Delayed_Payment: Represents the average number of payments delayed by a person       
Changed_Credit_Limit: Represents the percentage change in credit card limit            
Num_Credit_Inquiries: Represents the number of credit card inquiries        
Credit_Mix: Represents the classification of the mix of credits                      
Outstanding_Debt: Represents the remaining debt to be paid (in USD)                
Credit_Utilization_Ratio: Represents the utilization ratio of credit card        
Credit_History_Age: Represents the age of credit history of the person          
Payment_of_Min_Amount: Represents whether only the minimum amount was paid by the person          
Total_EMI_per_month: Represents the monthly EMI payments (in USD)             
Amount_invested_monthly: Represents the monthly amount invested by the customer (in USD)      
Payment_Behaviour: Represents the payment behavior of the customer (in USD)               
Monthly_Balance: Represents the monthly balance amount of the customer (in USD)                            

All of these features relate to a person's financial information like their credit history, investments, and bank accounts. Before data preprocessing, there are 100,000 records in the dataset.
The label for the dataset is the last column.
Credit Score: There are 3 bins for classification - Good, Bad, and Standard


In [143]:
# Imports:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

Read in the dataset from the CSV file and display the first 5 records. We set the option to show all the columns of the dataframe as it will cutout the middle by default when displaying it.

In [144]:
data = pd.read_csv('train.csv')
pd.set_option('display.max_columns', None)
data.head()

/var/folders/_k/wvslhc31643g32fz4j94qvyh0000gn/T/ipykernel_80323/4252536802.py:1: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('train.csv')


ID Customer_ID     Month           Name   Age          SSN Occupation  \
0  0x1602   CUS_0xd40   January  Aaron Maashoh    23  821-00-0265  Scientist   
1  0x1603   CUS_0xd40  February  Aaron Maashoh    23  821-00-0265  Scientist   
2  0x1604   CUS_0xd40     March  Aaron Maashoh  -500  821-00-0265  Scientist   
3  0x1605   CUS_0xd40     April  Aaron Maashoh    23  821-00-0265  Scientist   
4  0x1606   CUS_0xd40       May  Aaron Maashoh    23  821-00-0265  Scientist   

  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
0      19114.12            1824.843333                  3                4   
1      19114.12                    NaN                  3                4   
2      19114.12                    NaN                  3                4   
3      19114.12                    NaN                  3                4   
4      19114.12            1824.843333                  3                4   

   Interest_Rate Num_of_Loan  \
0              3           4   
1              3           4   
2              3           4   
3              3           4   
4              3           4   

                                        Type_of_Loan  Delay_from_due_date  \
0  Auto Loan, Credit-Builder Loan, Personal Loan,...                    3   
1  Auto Loan, Credit-Builder Loan, Personal Loan,...                   -1   
2  Auto Loan, Credit-Builder Loan, Personal Loan,...                    3   
3  Auto Loan, Credit-Builder Loan, Personal Loan,...                    5   
4  Auto Loan, Credit-Builder Loan, Personal Loan,...                    6   

  Num_of_Delayed_Payment Changed_Credit_Limit  Num_Credit_Inquiries  \
0                      7                11.27                   4.0   
1                    NaN                11.27                   4.0   
2                      7                    _                   4.0   
3                      4                 6.27                   4.0   
4                    NaN                11.27                   4.0   

  Credit_Mix Outstanding_Debt  Credit_Utilization_Ratio  \
0          _           809.98                 26.822620   
1       Good           809.98                 31.944960   
2       Good           809.98                 28.609352   
3       Good           809.98                 31.377862   
4       Good           809.98                 24.797347   

      Credit_History_Age Payment_of_Min_Amount  Total_EMI_per_month  \
0  22 Years and 1 Months                    No            49.574949   
1                    NaN                    No            49.574949   
2  22 Years and 3 Months                    No            49.574949   
3  22 Years and 4 Months                    No            49.574949   
4  22 Years and 5 Months                    No            49.574949   

  Amount_invested_monthly                 Payment_Behaviour  \
0       80.41529543900253   High_spent_Small_value_payments   
1      118.28022162236736    Low_spent_Large_value_payments   
2         81.699521264648   Low_spent_Medium_value_payments   
3       199.4580743910713    Low_spent_Small_value_payments   
4      41.420153086217326  High_spent_Medium_value_payments   

      Monthly_Balance Credit_Score  
0  312.49408867943663         Good  
1  284.62916249607184         Good  
2   331.2098628537912         Good  
3  223.45130972736786         Good  
4  341.48923103222177         Good

# Data Preproccesing:

### Observations
To begin, we went through the dataset and observed all features to see what we have to work with. As you can see below, there were many features that had a lot of NaN values, which we needed to clean up. At first, we thought about simply just dropping all the records with NaN values, but we realized that this was a good chunk of valuable data that we can use, so we had to explore further of what the best way to impute values were, which we will talk about later.

In [145]:
print(data.isnull().sum())

ID                              0
Customer_ID                     0
Month                           0
Name                         9985
Age                             0
SSN                             0
Occupation                      0
Annual_Income                   0
Monthly_Inhand_Salary       15002
Num_Bank_Accounts               0
Num_Credit_Card                 0
Interest_Rate                   0
Num_of_Loan                     0
Type_of_Loan                11408
Delay_from_due_date             0
Num_of_Delayed_Payment       7002
Changed_Credit_Limit            0
Num_Credit_Inquiries         1965
Credit_Mix                      0
Outstanding_Debt                0
Credit_Utilization_Ratio        0
Credit_History_Age           9030
Payment_of_Min_Amount           0
Total_EMI_per_month             0
Amount_invested_monthly      4479
Payment_Behaviour               0
Monthly_Balance              1200
Credit_Score                    0
dtype: int64


After observing the amount of null values, we wanted to see what data type each feature was, so we could determine if any changes were needed before we reach the modeling part. We first simply just used data.dtypes, but this didn't give us much information into the specific type because it gave us the type 'Object' for many of the features. Using the code below, we were able to print out a more specific type for each feature.

In [146]:
for col in data.columns:
    print(f'{col} : {pd.api.types.infer_dtype(data[col])}')

ID : string
Customer_ID : string
Month : string
Name : string
Age : string
SSN : string
Occupation : string
Annual_Income : string
Monthly_Inhand_Salary : floating
Num_Bank_Accounts : integer
Num_Credit_Card : integer
Interest_Rate : integer
Num_of_Loan : string
Type_of_Loan : string
Delay_from_due_date : integer
Num_of_Delayed_Payment : string
Changed_Credit_Limit : string
Num_Credit_Inquiries : floating
Credit_Mix : string
Outstanding_Debt : string
Credit_Utilization_Ratio : floating
Credit_History_Age : string
Payment_of_Min_Amount : string
Total_EMI_per_month : floating
Amount_invested_monthly : string
Payment_Behaviour : string
Monthly_Balance : mixed
Credit_Score : string


After examining these results, we quickly realized that many of the features that are supposed to be numerical in type are actually strings, which would cause issues during the modeling portion. Therefore, as we were cleaning data, we knew we had to check why this was the case, and change the type to be a float, so that it is usable in modeling. For instance, features like Number of Loans should be a float value, not a String. This was really strange to us, which we dove into as we looked at each feature later in this notebook.

Additionally, we were confused why column 26, which is Monthly Balance had a type for 'Mixed'. We wanted to see what this type of 'Mixed' meant, which is what the code below demonstrates. Apparently, this feature has some records with type String and some with type float. We explored this later when engineering and cleaning each feature.

In [147]:
column_data = data.iloc[ : , 26] 
print(column_data.apply(type).value_counts())

Monthly_Balance
<class 'str'>      97132
<class 'float'>     2868
Name: count, dtype: int64


## Data Cleaning

### Dropping Useless Features

We dropped features that we figured wouldn't have an impact on what a person's credit score is classified as. 
1. ID: This column simply is there to distinguish each record from each other. There is no correlation between the ID and if the certain customer has a good, bad, or standard credit score.
2. Name, SSN: These features simply just describe and identify the customer. They have no effect on the prediction.
3. Month: This shows which month the information is from, and is there to distinguish between the 8 records for each customer.

**Note: Customer_ID will be dropped later. We use it when we clean the data to group a customer with all their transactions/records**

As you can see below, these features have now been dropped.

In [148]:
# Drop useless features.
data.drop(['ID', 'Name', 'SSN', 'Month'], axis = 1, inplace = True)
data.head()

Customer_ID   Age Occupation Annual_Income  Monthly_Inhand_Salary  \
0   CUS_0xd40    23  Scientist      19114.12            1824.843333   
1   CUS_0xd40    23  Scientist      19114.12                    NaN   
2   CUS_0xd40  -500  Scientist      19114.12                    NaN   
3   CUS_0xd40    23  Scientist      19114.12                    NaN   
4   CUS_0xd40    23  Scientist      19114.12            1824.843333   

   Num_Bank_Accounts  Num_Credit_Card  Interest_Rate Num_of_Loan  \
0                  3                4              3           4   
1                  3                4              3           4   
2                  3                4              3           4   
3                  3                4              3           4   
4                  3                4              3           4   

                                        Type_of_Loan  Delay_from_due_date  \
0  Auto Loan, Credit-Builder Loan, Personal Loan,...                    3   
1  Auto Loan, Credit-Builder Loan, Personal Loan,...                   -1   
2  Auto Loan, Credit-Builder Loan, Personal Loan,...                    3   
3  Auto Loan, Credit-Builder Loan, Personal Loan,...                    5   
4  Auto Loan, Credit-Builder Loan, Personal Loan,...                    6   

  Num_of_Delayed_Payment Changed_Credit_Limit  Num_Credit_Inquiries  \
0                      7                11.27                   4.0   
1                    NaN                11.27                   4.0   
2                      7                    _                   4.0   
3                      4                 6.27                   4.0   
4                    NaN                11.27                   4.0   

  Credit_Mix Outstanding_Debt  Credit_Utilization_Ratio  \
0          _           809.98                 26.822620   
1       Good           809.98                 31.944960   
2       Good           809.98                 28.609352   
3       Good           809.98                 31.377862   
4       Good           809.98                 24.797347   

      Credit_History_Age Payment_of_Min_Amount  Total_EMI_per_month  \
0  22 Years and 1 Months                    No            49.574949   
1                    NaN                    No            49.574949   
2  22 Years and 3 Months                    No            49.574949   
3  22 Years and 4 Months                    No            49.574949   
4  22 Years and 5 Months                    No            49.574949   

  Amount_invested_monthly                 Payment_Behaviour  \
0       80.41529543900253   High_spent_Small_value_payments   
1      118.28022162236736    Low_spent_Large_value_payments   
2         81.699521264648   Low_spent_Medium_value_payments   
3       199.4580743910713    Low_spent_Small_value_payments   
4      41.420153086217326  High_spent_Medium_value_payments   

      Monthly_Balance Credit_Score  
0  312.49408867943663         Good  
1  284.62916249607184         Good  
2   331.2098628537912         Good  
3  223.45130972736786         Good  
4  341.48923103222177         Good

Before continuing on with the data preprocessing, we want to make sure that each customer has 8 records associated with them: the eight months (January - August) of their credit information. To do this, we are checking if there are any groups based on customer ID that are not 8. If they had groups that weren't 8, then the code below would print True, but it prints false. We successfully ensured that each customer has the same number of records before grouping them in the next steps.

In [149]:
(data.Customer_ID.value_counts() != 8).any()

False

While diving deep into each feature, we quickly realized that many records had really unreasonable outliers, like negative ages or values that were way too big to make sense. Therefore, we created this helper function that we used to clean many of the features within our dataset. 

What made this issue was unique had to do with the fact that each customer had 8 records associated with them, which is why we utilized mode. Many features like age, annual income, and number of bank accounts do not change that frequently month-to-month for a customer. Therefore, whenver we saw a customer with records that had age 23 for all 7 months and -500 for 1 month, which doesn't make sense at all, we replaced this outlier with the mode of the group (the 8 records for the particular customer). However, sometimes the mode doesn't exist, so in that case, we default to the mean. However, even the whole customer can be an outlier, rather than just 1 record. We noticed this when we saw that the mode value for a certain feature was out of the bounds of what we considered not an outlier. In this case, we dropped entire customer. To drop the entire customer, this helper function simply sets it to NaN, which gets dropped when we call .dropna later. This is a general explanation of what this helper function is supposed to do, and why we are cleaning the data in the way that we are.

In [150]:
def replace_or_drop(group, min, max):
    mode_series = group.mode()
    if mode_series.empty:
        replacement_value = group.mean()  # Fallback to mean if no mode exists
    else:
        mode_value = mode_series[0]
        if mode_value < min or mode_value > max:
            replacement_value = np.nan  # Set to NaN if mode is out of accepted range
        else:
            replacement_value = mode_value

    # Replace out-of-bounds and NaN values
    group = np.where((group < min) | (group > max) | group.isna(), replacement_value, group)
    
    # If replacement_value is NaN (due to mode being out of bounds), drop these by returning NaN
    if np.isnan(replacement_value):
        return pd.Series([np.nan] * len(group))
    return group

## Diving Into Features

### Age

The data type of age started out as a string. Some records had an age value that made sense, but the number had an underscore after it, which made it a string. For instance, if the age was 23, in the dataset, some records had this written as 23_. Therefore, we replaced all the underscore characters to be nothing. After this was taken care of, we converted the feature from string to float, which is what age is supposed to be.

Before cleaning, we printed out the description of the values in the feature. It is clear that this does not make sense at all and that there are really big outliers, skewing the data. The min for age is -500 and the max for age was 8698, leaving the mean to be an age over 100. These values are obviously outliers. We group the data by customer ID, which ensures that each customer is grouped together with all their records. The feature will then be cleaned, using the helper function I described above. We used 0 and 56 as the min and max to determine outliers as 56 was the highest age we had besides the massive outlier. 0 was the lower bound to get rid of any negative ages that may be in the dataset.

After transforming the data with replace_or_drop, we will drop all the records with NaN for age, which means the entire customer was an outlier. However, you can see that no records were dropped because of NaN values as we remained with 100,000 records after the transformation.

In [151]:
data['Age'] = data['Age'].str.replace('_', '')
data['Age'] = data['Age'].replace("", np.nan)
data['Age'] = data['Age'].astype(float)
print('Before Cleaning:')
print(data['Age'].describe())

grouped_data = data.groupby('Customer_ID')
data['Age'] = grouped_data['Age'].transform(lambda g: replace_or_drop(g, 0, 56))

data.dropna(subset=['Age'], inplace = True)
data.reset_index(drop = True, inplace = True)
print('After Cleaning:')
print(data['Age'].describe())

Before Cleaning:
count    100000.000000
mean        110.649700
std         686.244717
min        -500.000000
25%          24.000000
50%          33.000000
75%          42.000000
max        8698.000000
Name: Age, dtype: float64


After Cleaning:
count    100000.000000
mean         33.311180
std          10.764783
min          14.000000
25%          24.000000
50%          33.000000
75%          42.000000
max          56.000000
Name: Age, dtype: float64


### Annual Income

The data type of annual income started out as a string. Some records had an income value that made sense, but the number had an underscore after it, which made it a string (similar to the age feature). Therefore, we replaced all the underscore characters to be nothing. After this was taken care of, we converted the feature from string to float, which is what annual income is supposed to be.

Before cleaning, we printed out the description of the values in the feature. We didn't notice any outliers and there are no records that have NaN for annual income, so that should be it for cleaning up the annual income feature.

In [152]:
data['Annual_Income'] = data['Annual_Income'].str.replace('_', '')
data['Annual_Income'] = data['Annual_Income'].replace("", np.nan)
data['Annual_Income'] = data['Annual_Income'].astype(float)
print(data['Annual_Income'].describe())

# grouped_data = data.groupby('Customer_ID')
# data['Annual_Income'] = grouped_data['Annual_Income'].transform(lambda g: replace_or_drop(g, 0, 2.5e+07))

# # Drop all rows where 'Num_Bank_Accounts' is NaN, these rows are from groups with a negative mode
# data.dropna(subset=['Annual_Income'], inplace = True)
# data.reset_index(drop = True, inplace = True)

# print(data['Annual_Income'].describe())
# Annual Income now looks good.

count    1.000000e+05
mean     1.764157e+05
std      1.429618e+06
min      7.005930e+03
25%      1.945750e+04
50%      3.757861e+04
75%      7.279092e+04
max      2.419806e+07
Name: Annual_Income, dtype: float64


### Monthly Inhand Salary

Montly Inhand Salary was already a float value, so we did not need to change that for this feature. There were some NaN values, however, so we wanted to replace those values with the mode of the specific customer's monthly inhand salary. 

If there are still NaN values left after the grouping and calling replace_or_drop, this means that the group itself had a mode that was a big outlier, so we want to drop this customer as a whole.

Normal range for this feature: 0 - 15205

As you can see, the count value went up from 84,998 to 100,000 because the records with NaN monthly inhand salaries are now filled with the customer's monthly inhand salary mode.

In [153]:
print(data['Monthly_Inhand_Salary'].describe())

grouped_data = data.groupby('Customer_ID')
data['Monthly_Inhand_Salary'] = grouped_data['Monthly_Inhand_Salary'].transform(lambda g: replace_or_drop(g, 0, 15205))

data.dropna(subset=['Monthly_Inhand_Salary'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data['Monthly_Inhand_Salary'].describe())

count    84998.000000
mean      4194.170850
std       3183.686167
min        303.645417
25%       1625.568229
50%       3093.745000
75%       5957.448333
max      15204.633333
Name: Monthly_Inhand_Salary, dtype: float64
count    100000.000000
mean       4198.350924
std        3187.402015
min         303.645417
25%        1626.761667
50%        3095.978333
75%        5961.637500
max       15204.633333
Name: Monthly_Inhand_Salary, dtype: float64


### Number of Bank Accounts

We first had to change the type of number of bank accounts to be a float instead of a string. Similar to the previous features, there were records with outlier values (negative numbers and numbers extremely high like 1798). We needed to replace these values with the mode of the customer's number of bank accounts. However, if the mode for the customer was still an outlier, then we dropped the entire customer (all 8 rows) from the dataset. 

Normal range for this feature: 0 - 10

As you can see, there were some records that got dropped completely.

In [154]:
data['Num_Bank_Accounts'] = data['Num_Bank_Accounts'].astype(float)
print(data['Num_Bank_Accounts'].describe())

grouped_data = data.groupby('Customer_ID')
data['Num_Bank_Accounts'] = grouped_data['Num_Bank_Accounts'].transform(lambda g: replace_or_drop(g, 0, 10))

data.dropna(subset = ['Num_Bank_Accounts'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data['Num_Bank_Accounts'].describe())

count    100000.000000
mean         17.091280
std         117.404834
min          -1.000000
25%           3.000000
50%           6.000000
75%           7.000000
max        1798.000000
Name: Num_Bank_Accounts, dtype: float64
count    99968.000000
mean         5.370469
std          2.591730
min          0.000000
25%          3.000000
50%          5.000000
75%          7.000000
max         10.000000
Name: Num_Bank_Accounts, dtype: float64


### Number of Credit Cards

We first had to change the type of number of credit cards to be a float instead of a string. Similar to the previous features, there were records with outlier values (really extreme values like 1499). We needed to replace these values with the mode of the customer's number of bank accounts. However, if the mode for the customer was still an outlier, then we dropped the entire customer (all 8 rows) from the dataset. 

Normal range for this feature: 0 - 10

As you can see, there were some records that got dropped completely.

In [155]:
data['Num_Credit_Card'] = data['Num_Credit_Card'].astype(float)
print(data['Num_Credit_Card'].describe())

grouped_data = data.groupby('Customer_ID')
data['Num_Credit_Card'] = grouped_data['Num_Credit_Card'].transform(lambda g: replace_or_drop(g, 0, 10))

data.dropna(subset=['Num_Credit_Card'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data["Num_Credit_Card"].describe())

count    99968.000000
mean        22.478383
std        129.076914
min          0.000000
25%          4.000000
50%          5.000000
75%          7.000000
max       1499.000000
Name: Num_Credit_Card, dtype: float64
count    99928.000000
mean         5.531663
std          2.064943
min          0.000000
25%          4.000000
50%          5.000000
75%          7.000000
max         10.000000
Name: Num_Credit_Card, dtype: float64


### Interest Rate

We first had to change the type of interest to be a float instead of a string. Similar to the previous features, there were records with outlier values (really extreme values like 5797). It doesn't make sense to have an interest rate that high, so this was considered an outlier/fake data. We needed to replace these values with the mode of the customer's number of bank accounts. However, if the mode for the customer was still an outlier, then we dropped the entire customer (all 8 rows) from the dataset. 

Normal range for this feature: 0 - 36

We didn't know what values were reasonable for the credit card interest rate, but after doing some research, we found that the max credit card rate right now is 36%, which is why we considered 0 to 36 as the range of possible values.

No records were dropped because of this feature.

In [156]:
data['Interest_Rate'] = data['Interest_Rate'].astype(float)
print(data["Interest_Rate"].describe())

grouped_data = data.groupby('Customer_ID')
data['Interest_Rate'] = grouped_data['Interest_Rate'].transform(lambda g: replace_or_drop(g, 0, 36))

data.dropna(subset=['Interest_Rate'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data["Interest_Rate"].describe())

count    99928.000000
mean        72.472400
std        466.454174
min          1.000000
25%          8.000000
50%         13.000000
75%         20.000000
max       5797.000000
Name: Interest_Rate, dtype: float64
count    99928.000000
mean        14.532543
std          8.741033
min          1.000000
25%          7.000000
50%         13.000000
75%         20.000000
max         34.000000
Name: Interest_Rate, dtype: float64


### Number of Loans

We first had to change the type of number of loans to be a float instead of a string. This feature also had the issue of having underscores after the numbers for some records, which is something we needed to fix. Similar to the previous features, there were records with outlier values (negative numbers and numbers extremely high like 1496). We needed to replace these values with the mode of the customer's number of bank accounts. However, if the mode for the customer was still an outlier, then we dropped the entire customer (all 8 rows) from the dataset. 

Normal range for this feature: 0 - 10

There were no records dropped when cleaning up number of loans.

In [157]:
data['Num_of_Loan'] = data['Num_of_Loan'].str.replace('_', '')
data['Num_of_Loan'] = data['Num_of_Loan'].replace("", np.nan)
data['Num_of_Loan'] = data['Num_of_Loan'].astype(float)
print(data['Num_of_Loan'].describe())

grouped_data = data.groupby('Customer_ID')
data['Num_of_Loan'] = grouped_data['Num_of_Loan'].transform(lambda g: replace_or_drop(g, 0, 10))

data.dropna(subset=['Num_of_Loan'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data['Num_of_Loan'].describe())

count    99928.00000
mean         3.01366
std         62.66703
min       -100.00000
25%          1.00000
50%          3.00000
75%          5.00000
max       1496.00000
Name: Num_of_Loan, dtype: float64
count    99928.000000
mean         3.532784
std          2.446580
min          0.000000
25%          2.000000
50%          3.000000
75%          5.000000
max          9.000000
Name: Num_of_Loan, dtype: float64


### Delay from Due Date

We first had to change the delay from due date feature to be a float instead of a string. There were records with outlier values (negative values), which simply do not make sense. We needed to replace these values with the mode of the customer's number of bank accounts. However, if the mode for the customer was still an outlier, then we dropped the entire customer (all 8 rows) from the dataset. 

Normal range for this feature: 0 - 67

As you can see, there were some records that got dropped completely. We went from 99928 to 99888 records.

In [158]:
data['Delay_from_due_date'] = data['Delay_from_due_date'].astype(float)
print(data["Delay_from_due_date"].describe())

grouped_data = data.groupby('Customer_ID')
data['Delay_from_due_date'] = grouped_data['Delay_from_due_date'].transform(lambda g: replace_or_drop(g, 0, 67))

data.dropna(subset=['Delay_from_due_date'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data["Delay_from_due_date"].describe())

count    99928.000000
mean        21.064566
std         14.855445
min         -5.000000
25%         10.000000
50%         18.000000
75%         28.000000
max         67.000000
Name: Delay_from_due_date, dtype: float64
count    99888.000000
mean        21.093885
std         14.822143
min          0.000000
25%         10.000000
50%         18.000000
75%         28.000000
max         67.000000
Name: Delay_from_due_date, dtype: float64


### Number of Delayed Payment

We first had to change the type of number of delayed payments to be a float instead of a string. After printing out the description of the features before we cleaned, we noticed the negative values and extreme values like 4397, which don't really make sense. We needed to replace these values with the mode of the customer's number of bank accounts. However, if the mode for the customer was still an outlier, then we dropped the entire customer (all 8 rows) from the dataset. 

Normal range for this feature: 0 - 31

The count variable increased, which means we were able to fill in some of the NaN values with values that make sense for the particular customer.

In [159]:
data['Num_of_Delayed_Payment'] = data['Num_of_Delayed_Payment'].str.replace('_', '')
data['Num_of_Delayed_Payment'] = data['Num_of_Delayed_Payment'].replace("", np.nan)
data['Num_of_Delayed_Payment'] = data['Num_of_Delayed_Payment'].astype(float)
print(data['Num_of_Delayed_Payment'].describe())

grouped_data = data.groupby('Customer_ID')
data['Num_of_Delayed_Payment'] = grouped_data['Num_of_Delayed_Payment'].transform(lambda g: replace_or_drop(g, 0, 31))

data.dropna(subset=['Num_of_Delayed_Payment'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data['Num_of_Delayed_Payment'].describe())

count    92894.000000
mean        30.912244
std        225.911493
min         -3.000000
25%          9.000000
50%         14.000000
75%         18.000000
max       4397.000000
Name: Num_of_Delayed_Payment, dtype: float64
count    99776.000000
mean        13.342397
std          6.258616
min          0.000000
25%          9.000000
50%         14.000000
75%         18.000000
max         28.000000
Name: Num_of_Delayed_Payment, dtype: float64


### Changed Credit Limit

We first had to change the type of changed credit limit to be a float instead of a string. This had the same issue with the underscore, so we needed to fix that first. We didn't find any outliers in this feature as negative values are allowed since this means a decrease in changed credit limit.  

Normal range for this feature: -6.5 - 37

As you can see, there were some records that now don't have NaN values for changed credit limit.

In [160]:
data['Changed_Credit_Limit'] = data['Changed_Credit_Limit'].str.replace('_', '')
data['Changed_Credit_Limit'] = data['Changed_Credit_Limit'].replace("", np.nan)
data['Changed_Credit_Limit'] = data['Changed_Credit_Limit'].astype(float)
print(data['Changed_Credit_Limit'].describe())

grouped_data = data.groupby('Customer_ID')
data['Changed_Credit_Limit'] = grouped_data['Changed_Credit_Limit'].transform(lambda g: replace_or_drop(g, -6.5, 37))

data.dropna(subset=['Changed_Credit_Limit'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data['Changed_Credit_Limit'].describe())

count    97693.000000
mean        10.393692
std          6.791314
min         -6.490000
25%          5.340000
50%          9.410000
75%         14.880000
max         36.970000
Name: Changed_Credit_Limit, dtype: float64
count    99776.000000
mean        10.393665
std          6.785547
min         -6.490000
25%          5.340000
50%          9.400000
75%         14.870000
max         36.970000
Name: Changed_Credit_Limit, dtype: float64


### Number of Credit Inquiries

Number of Credit Inquiries was already a float so there was no need to convert the data type of this feature. Furthermore, similar to the other features, we want to replace any NaN values and outlier values with the mode for number of credit inquiries for each specific customer. In the dataset, the value 2597 seemed unrealistic and fake for number of credit inquiries, so we decided that this was an outlier value.

Normal Value: 0 - 12

Since there were customers that had an outlier mode all together, this customer was dropped. As you can see, we lost some records due to the mode being an outlier for the customer.

In [161]:
print(data['Num_Credit_Inquiries'].describe())

grouped_data = data.groupby('Customer_ID')
data['Num_Credit_Inquiries'] = grouped_data['Num_Credit_Inquiries'].transform(lambda g: replace_or_drop(g, 0, 12))

data.dropna(subset=['Num_Credit_Inquiries'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data['Num_Credit_Inquiries'].describe())

count    97819.000000
mean        27.788967
std        193.342315
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max       2597.000000
Name: Num_Credit_Inquiries, dtype: float64
count    96056.000000
mean         5.433809
std          3.500103
min          0.000000
25%          3.000000
50%          5.000000
75%          8.000000
max         12.000000
Name: Num_Credit_Inquiries, dtype: float64


### Outstanding Debt

Outstanding Debt had the same issue with the underscores after numbers, so that was something that needed to be cleaned, which allowed us to then convert the values into floating values. We then transform the column to replace any outliers and NaN with the same logic we have been using. 

Normal Value: 0 - 5000

In [162]:
data['Outstanding_Debt'] = data['Outstanding_Debt'].str.replace('_', '')
data['Outstanding_Debt'] = data['Outstanding_Debt'].replace("", np.nan)
data['Outstanding_Debt'] = data['Outstanding_Debt'].astype(float)
print(data['Outstanding_Debt'].describe())

# grouped_data = data.groupby('Customer_ID')
# data['Outstanding_Debt'] = grouped_data['Outstanding_Debt'].transform(lambda g: replace_or_drop(g, 0, 5000))

# data.dropna(subset=['Outstanding_Debt'], inplace = True)
# data.reset_index(drop = True, inplace = True)

# print(data['Outstanding_Debt'].describe())
# Outstanding debt looks good. Although we have values near 5000 when the median is near 1170, it is reasonable that people have this much debt.

count    96056.000000
mean      1374.910821
std       1126.104035
min          0.230000
25%        550.220000
50%       1123.230000
75%       1811.960000
max       4998.070000
Name: Outstanding_Debt, dtype: float64


### Credit Utilization Ratio

Credit Utilization Ratio was already a float so there was no need to convert the data type of this feature. Furthermore, similar to the other features, we want to replace any NaN values and outlier values with the mode for credit utilization ratio for each specific customer.

Normal Value: 20 - 50

Since there were customers that had an outlier mode all together, this customer was dropped. However, there were no records dropped when cleaning this feature.

In [163]:
print(data['Credit_Utilization_Ratio'].describe())

grouped_data = data.groupby('Customer_ID')
data['Credit_Utilization_Ratio'] = grouped_data['Credit_Utilization_Ratio'].transform(lambda g: replace_or_drop(g, 20, 50))

data.dropna(subset=['Credit_Utilization_Ratio'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data['Credit_Utilization_Ratio'].describe())

count    96056.000000
mean        32.308582
std          5.119054
min         20.000000
25%         28.075516
50%         32.329881
75%         36.518995
max         50.000000
Name: Credit_Utilization_Ratio, dtype: float64
count    96056.000000
mean        32.308384
std          5.118738
min         20.000000
25%         28.075516
50%         32.329582
75%         36.518741
max         49.564519
Name: Credit_Utilization_Ratio, dtype: float64


### Total EMI Per Month

In [164]:
# print(data['Total_EMI_per_month'].describe())

# grouped_data = data.groupby('Customer_ID')
# data['Total_EMI_per_month'] = grouped_data['Total_EMI_per_month'].transform(lambda g: replace_or_drop(g, 0, 90000))

# data.dropna(subset=['Total_EMI_per_month'], inplace = True)
# data.reset_index(drop = True, inplace = True)

# print(data['Total_EMI_per_month'].describe())

### Amount Invested Monthly

The biggest thing we needed to change for amount invested monthly was changing the data type of the feature. Furthermore, we needed to impute the amount invested monthly for a customer into the NaN values as this feature had a lot of NaN values within its column.

In [165]:
data['Amount_invested_monthly'] = data['Amount_invested_monthly'].str.replace('_', '')
data['Amount_invested_monthly'] = data['Amount_invested_monthly'].replace("", np.nan)
data['Amount_invested_monthly'] = data['Amount_invested_monthly'].astype(float)
print(data['Amount_invested_monthly'].describe())

grouped_data = data.groupby('Customer_ID')
data['Amount_invested_monthly'] = grouped_data['Amount_invested_monthly'].transform(lambda g: replace_or_drop(g, 0, 10000))

data.dropna(subset=['Amount_invested_monthly'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data['Amount_invested_monthly'].describe())

count    91752.000000
mean       639.769567
std       2043.966987
min          0.000000
25%         75.240647
50%        137.184427
75%        268.975996
max      10000.000000
Name: Amount_invested_monthly, dtype: float64
count    96056.000000
mean       631.070128
std       2037.974177
min          0.000000
25%         71.701968
50%        131.812904
75%        259.900991
max      10000.000000
Name: Amount_invested_monthly, dtype: float64


### Monthly Balance

This feature first needed to be fixed by removing the underscores next to any numbers that got added into the dataset. We, then, could transform the feature into a float instead of a string. This feature had the biggest outliers we have seen. A monthly balance of -3.33e26 is so unreasonable and is fake data, so we had to ensure we got replaced or got rid of all the records with values this extreme.

Normal value for monthly balance: 0 - 1580

We were able to recover some records that had NaN values originally that we replaced with the mode of the monthly balance for a specific customer.

In [166]:
data['Monthly_Balance'] = data['Monthly_Balance'].str.replace('_', '')
data['Monthly_Balance'] = data['Monthly_Balance'].replace("", np.nan)
data['Monthly_Balance'] = data['Monthly_Balance'].astype(float)
print(data['Monthly_Balance'].describe())

grouped_data = data.groupby('Customer_ID')
data['Monthly_Balance'] = grouped_data['Monthly_Balance'].transform(lambda g: replace_or_drop(g, 0, 1.58e+03))

data.dropna(subset=['Monthly_Balance'], inplace = True)
data.reset_index(drop = True, inplace = True)

print(data['Monthly_Balance'].describe())

count    9.326600e+04
mean    -2.859206e+22
std      3.087065e+24
min     -3.333333e+26
25%      2.717432e+02
50%      3.395629e+02
75%      4.756614e+02
max      1.602041e+03
Name: Monthly_Balance, dtype: float64
count    94328.000000
mean       403.757802
std        215.963790
min          0.007760
25%        270.287890
50%        338.317664
75%        473.762700
max       1576.288935
Name: Monthly_Balance, dtype: float64


## Feature Engineering

### Credit History Age

It was hard to use this feature as it came in the dataset because it was a string that said how many years and months. We figured that we could simply convert this feature to be the total number of months for credit history age, instead of being in terms of both years and months. We have this function called years_months_to_months, which essentially parses through the string, and converts the total number of months from it. Each record's credit history age is then replaced by the number of months instead. 

Once we engineer this feature to be total number of months, we then transform it the same way we did with other features to get rid of any NaN and outlier values.

Normal number of months: 0 - 405

In [167]:
# Filter the dataframe to keep only rows where 'Credit_History_Age' is a string
data = data[data['Credit_History_Age'].apply(lambda x: isinstance(x, str))]

# Now apply the conversion function
def years_months_to_months(s):
    if s == "NA":
        return np.nan
    parts = s.split(' ')
    years = int(parts[0])
    months = int(parts[3])
    total_months = years * 12 + months
    return total_months

data['Credit_History_Age'] = data['Credit_History_Age'].apply(years_months_to_months).astype(float)

grouped_data = data.groupby('Customer_ID')
data['Credit_History_Age'] = grouped_data['Credit_History_Age'].transform(lambda g: replace_or_drop(g, 0, 405))

data.dropna(subset=['Credit_History_Age'], inplace = True)
data.reset_index(drop = True, inplace = True)

data.head()
print(data['Credit_History_Age'].describe())
# TOOOODOOOOOOOOOOO!!!

/var/folders/_k/wvslhc31643g32fz4j94qvyh0000gn/T/ipykernel_80323/379135883.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Credit_History_Age'] = data['Credit_History_Age'].apply(years_months_to_months).astype(float)


count    85819.000000
mean       224.724257
std         99.196315
min          1.000000
25%        150.000000
50%        223.000000
75%        306.000000
max        404.000000
Name: Credit_History_Age, dtype: float64


### Occupation

Occupation is a categorical variable that takes in values of different jobs like "Lawyer", "Teacher", "Engineer", etc. We want to be able to use this feature in our machine learning model, and there are some models that require numbers for inputs. Therefore, we can one hot encode this feature to represent a categorical variable as a numerical value within the dataset. Some records have no occupation in which we set the title to Unemployed.

In [168]:
data['Occupation'].replace("_______", "Unemployed", inplace=True)
print(data['Occupation'].value_counts())

# One hot encode:
occupation_dummies = pd.get_dummies(data['Occupation'], prefix = 'Occupation')
data = pd.concat([data, occupation_dummies], axis = 1)
data.drop('Occupation', axis = 1, inplace = True)

data.head()

Occupation
Unemployed       6076
Lawyer           5589
Architect        5464
Engineer         5436
Scientist        5414
Teacher          5388
Media_Manager    5358
Mechanic         5336
Entrepreneur     5335
Accountant       5322
Journalist       5289
Developer        5288
Doctor           5278
Manager          5152
Musician         5101
Writer           4993
Name: count, dtype: int64


/var/folders/_k/wvslhc31643g32fz4j94qvyh0000gn/T/ipykernel_80323/775808291.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Occupation'].replace("_______", "Unemployed", inplace=True)


Customer_ID   Age  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
0   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
1   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
2   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
3   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
4   CUS_0xd40  23.0       19114.12            1824.843333                3.0   

   Num_Credit_Card  Interest_Rate  Num_of_Loan  \
0              4.0            3.0          4.0   
1              4.0            3.0          4.0   
2              4.0            3.0          4.0   
3              4.0            3.0          4.0   
4              4.0            3.0          4.0   

                                        Type_of_Loan  Delay_from_due_date  \
0  Auto Loan, Credit-Builder Loan, Personal Loan,...                  3.0   
1  Auto Loan, Credit-Builder Loan, Personal Loan,...                  3.0   
2  Auto Loan, Credit-Builder Loan, Personal Loan,...                  5.0   
3  Auto Loan, Credit-Builder Loan, Personal Loan,...                  6.0   
4  Auto Loan, Credit-Builder Loan, Personal Loan,...                  8.0   

   Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
0                     7.0                 11.27                   4.0   
1                     7.0                 11.27                   4.0   
2                     4.0                  6.27                   4.0   
3                     4.0                 11.27                   4.0   
4                     4.0                  9.27                   4.0   

  Credit_Mix  Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
0          _            809.98                 26.822620               265.0   
1       Good            809.98                 28.609352               267.0   
2       Good            809.98                 31.377862               268.0   
3       Good            809.98                 24.797347               269.0   
4       Good            809.98                 27.262259               270.0   

  Payment_of_Min_Amount  Total_EMI_per_month  Amount_invested_monthly  \
0                    No            49.574949                80.415295   
1                    No            49.574949                81.699521   
2                    No            49.574949               199.458074   
3                    No            49.574949                41.420153   
4                    No            49.574949                62.430172   

                  Payment_Behaviour  Monthly_Balance Credit_Score  \
0   High_spent_Small_value_payments       312.494089         Good   
1   Low_spent_Medium_value_payments       331.209863         Good   
2    Low_spent_Small_value_payments       223.451310         Good   
3  High_spent_Medium_value_payments       341.489231         Good   
4                            !@9#%8       340.479212         Good   

   Occupation_Accountant  Occupation_Architect  Occupation_Developer  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   Occupation_Doctor  Occupation_Engineer  Occupation_Entrepreneur  \
0              False                False                    False   
1              False                False                    False   
2              False                False                    False   
3              False                False                    False   
4              False                False                    False   

   Occupation_Journalist  Occupation_Lawyer  Occupation_Manager  \
0                  False              False               Fals

### Type of Loan

This is another categorical feature that needs to be one hot encoded, so that we can use it in our machine learning models. This feature was slightly different because each record could have multiple types of loans. For instance, some records for this feature might say 'Auto Loan, Home Equity Loan, Personal Loan' which means that this customer has all of these types of loans. We needed to be able to parse through the list and ensure that each loan was marked as true if the customer had it listed in their type of loan.

In [169]:
all_loans = set()
data['Type_of_Loan'].dropna().str.split(',').apply(lambda x: [all_loans.add(loan.strip().replace('and ', '').replace('And ', '')) for loan in x])

# Function to create one-hot encodings
def encode_loans(row):
    if pd.isna(row):
        return pd.Series({loan: 0 for loan in all_loans})
    loans = [loan.strip().replace(' and', '').replace('And', '') for loan in row.split(',')]
    return pd.Series({loan: (1 if loan in loans else 0) for loan in all_loans})

# Apply the function across the Type_of_Loan column
loan_encoded = data['Type_of_Loan'].apply(encode_loans)

# Join the one-hot encoded DataFrame with the original DataFrame
data = data.join(loan_encoded)

# Optionally, drop the original 'Type_of_Loan' column if no longer needed
data.drop('Type_of_Loan', axis=1, inplace=True)

# Print the modified DataFrame
data.head()

# ********* TODO ************* #

Customer_ID   Age  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
0   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
1   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
2   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
3   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
4   CUS_0xd40  23.0       19114.12            1824.843333                3.0   

   Num_Credit_Card  Interest_Rate  Num_of_Loan  Delay_from_due_date  \
0              4.0            3.0          4.0                  3.0   
1              4.0            3.0          4.0                  3.0   
2              4.0            3.0          4.0                  5.0   
3              4.0            3.0          4.0                  6.0   
4              4.0            3.0          4.0                  8.0   

   Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
0                     7.0                 11.27                   4.0   
1                     7.0                 11.27                   4.0   
2                     4.0                  6.27                   4.0   
3                     4.0                 11.27                   4.0   
4                     4.0                  9.27                   4.0   

  Credit_Mix  Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
0          _            809.98                 26.822620               265.0   
1       Good            809.98                 28.609352               267.0   
2       Good            809.98                 31.377862               268.0   
3       Good            809.98                 24.797347               269.0   
4       Good            809.98                 27.262259               270.0   

  Payment_of_Min_Amount  Total_EMI_per_month  Amount_invested_monthly  \
0                    No            49.574949                80.415295   
1                    No            49.574949                81.699521   
2                    No            49.574949               199.458074   
3                    No            49.574949                41.420153   
4                    No            49.574949                62.430172   

                  Payment_Behaviour  Monthly_Balance Credit_Score  \
0   High_spent_Small_value_payments       312.494089         Good   
1   Low_spent_Medium_value_payments       331.209863         Good   
2    Low_spent_Small_value_payments       223.451310         Good   
3  High_spent_Medium_value_payments       341.489231         Good   
4                            !@9#%8       340.479212         Good   

   Occupation_Accountant  Occupation_Architect  Occupation_Developer  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   Occupation_Doctor  Occupation_Engineer  Occupation_Entrepreneur  \
0              False                False                    False   
1              False                False                    False   
2              False                False                    False   
3              False                False                    False   
4              False                False                    False   

   Occupation_Journalist  Occupation_Lawyer  Occupation_Manager  \
0                  False              False               False   
1                  False              False               False   
2                  False              False               False   
3                  False              False               False   
4                  False              False               False   

   Occupation_Mechanic  Occupation_Media_Manager  Occupation_Mu

In [170]:
# Function to replace invalid strings with the mode if valid
def replace_invalid_with_mode(group, invalid_str):
    mode_value = group.mode()
    if mode_value.empty or (mode_value[0] == invalid_str):
        return pd.Series([None] * len(group))  # Mark group for deletion if no valid mode or mode is invalid
    return group.replace(invalid_str, mode_value[0])

### Credit Mix

Credit Mix is another categorical feature that requires one hot encoding. Before we one hot encoded, we ensured that there were no NaN values by replacing any of these values with the most occuring credit mix for a specific customer. After, we one hot encode this feature, which has Good, Bad, or Standard.

In [171]:
print(data['Credit_Mix'].value_counts())

grouped_data = data.groupby('Customer_ID')
data['Credit_Mix'] = grouped_data['Credit_Mix'].transform(lambda x: replace_invalid_with_mode(x, '_'))

data.dropna(subset=['Credit_Mix'], inplace = True)

print(data['Credit_Mix'].value_counts())

# One hot encode:
occupation_dummies = pd.get_dummies(data['Credit_Mix'], prefix = 'Credit Mix')
data = pd.concat([data, occupation_dummies], axis = 1)
data.drop('Credit_Mix', axis = 1, inplace = True)

data.head()

Credit_Mix
Standard    31822
Good        21638
_           17306
Bad         15053
Name: count, dtype: int64
Credit_Mix
Standard    39247
Good        26505
Bad         18384
Name: count, dtype: int64


Customer_ID   Age  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
0   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
1   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
2   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
3   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
4   CUS_0xd40  23.0       19114.12            1824.843333                3.0   

   Num_Credit_Card  Interest_Rate  Num_of_Loan  Delay_from_due_date  \
0              4.0            3.0          4.0                  3.0   
1              4.0            3.0          4.0                  3.0   
2              4.0            3.0          4.0                  5.0   
3              4.0            3.0          4.0                  6.0   
4              4.0            3.0          4.0                  8.0   

   Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
0                     7.0                 11.27                   4.0   
1                     7.0                 11.27                   4.0   
2                     4.0                  6.27                   4.0   
3                     4.0                 11.27                   4.0   
4                     4.0                  9.27                   4.0   

   Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
0            809.98                 26.822620               265.0   
1            809.98                 28.609352               267.0   
2            809.98                 31.377862               268.0   
3            809.98                 24.797347               269.0   
4            809.98                 27.262259               270.0   

  Payment_of_Min_Amount  Total_EMI_per_month  Amount_invested_monthly  \
0                    No            49.574949                80.415295   
1                    No            49.574949                81.699521   
2                    No            49.574949               199.458074   
3                    No            49.574949                41.420153   
4                    No            49.574949                62.430172   

                  Payment_Behaviour  Monthly_Balance Credit_Score  \
0   High_spent_Small_value_payments       312.494089         Good   
1   Low_spent_Medium_value_payments       331.209863         Good   
2    Low_spent_Small_value_payments       223.451310         Good   
3  High_spent_Medium_value_payments       341.489231         Good   
4                            !@9#%8       340.479212         Good   

   Occupation_Accountant  Occupation_Architect  Occupation_Developer  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   Occupation_Doctor  Occupation_Engineer  Occupation_Entrepreneur  \
0              False                False                    False   
1              False                False                    False   
2              False                False                    False   
3              False                False                    False   
4              False                False                    False   

   Occupation_Journalist  Occupation_Lawyer  Occupation_Manager  \
0                  False              False               False   
1                  False              False               False   
2                  False              False               False   
3                  False              False               False   
4                  False              False               False   

   Occupation_Mechanic  Occupation_Media_Manager  Occupation_Musician  \
0                False                     False        

### Payment of Minimum Amount

Payment of Minimum Amount is another categorical feature that requires one hot encoding. Before we one hot encoded, we ensured that there were no NM values (NaN values in this case) by replacing any of these values with the most occuring credit mix for a specific customer. After, we one hot encode this feature, which is either Yes or No.

In [172]:
print(data['Payment_of_Min_Amount'].value_counts())

grouped_data = data.groupby('Customer_ID')
data['Payment_of_Min_Amount'] = grouped_data['Payment_of_Min_Amount'].transform(lambda x: replace_invalid_with_mode(x, 'NM'))

data.dropna(subset=['Payment_of_Min_Amount'], inplace = True)

print(data['Payment_of_Min_Amount'].value_counts())

# One hot encode:
occupation_dummies = pd.get_dummies(data['Payment_of_Min_Amount'], prefix = 'Payment_of_Min_Amount')
data = pd.concat([data, occupation_dummies], axis = 1)
data.drop('Payment_of_Min_Amount', axis = 1, inplace = True)

data.head()

Payment_of_Min_Amount
Yes    42876
No     31162
NM     10098
Name: count, dtype: int64
Payment_of_Min_Amount
Yes    48136
No     35142
Name: count, dtype: int64


Customer_ID   Age  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
0   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
1   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
2   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
3   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
4   CUS_0xd40  23.0       19114.12            1824.843333                3.0   

   Num_Credit_Card  Interest_Rate  Num_of_Loan  Delay_from_due_date  \
0              4.0            3.0          4.0                  3.0   
1              4.0            3.0          4.0                  3.0   
2              4.0            3.0          4.0                  5.0   
3              4.0            3.0          4.0                  6.0   
4              4.0            3.0          4.0                  8.0   

   Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
0                     7.0                 11.27                   4.0   
1                     7.0                 11.27                   4.0   
2                     4.0                  6.27                   4.0   
3                     4.0                 11.27                   4.0   
4                     4.0                  9.27                   4.0   

   Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
0            809.98                 26.822620               265.0   
1            809.98                 28.609352               267.0   
2            809.98                 31.377862               268.0   
3            809.98                 24.797347               269.0   
4            809.98                 27.262259               270.0   

   Total_EMI_per_month  Amount_invested_monthly  \
0            49.574949                80.415295   
1            49.574949                81.699521   
2            49.574949               199.458074   
3            49.574949                41.420153   
4            49.574949                62.430172   

                  Payment_Behaviour  Monthly_Balance Credit_Score  \
0   High_spent_Small_value_payments       312.494089         Good   
1   Low_spent_Medium_value_payments       331.209863         Good   
2    Low_spent_Small_value_payments       223.451310         Good   
3  High_spent_Medium_value_payments       341.489231         Good   
4                            !@9#%8       340.479212         Good   

   Occupation_Accountant  Occupation_Architect  Occupation_Developer  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   Occupation_Doctor  Occupation_Engineer  Occupation_Entrepreneur  \
0              False                False                    False   
1              False                False                    False   
2              False                False                    False   
3              False                False                    False   
4              False                False                    False   

   Occupation_Journalist  Occupation_Lawyer  Occupation_Manager  \
0                  False              False               False   
1                  False              False               False   
2                  False              False               False   
3                  False              False               False   
4                  False              False               False   

   Occupation_Mechanic  Occupation_Media_Manager  Occupation_Musician  \
0                False                     False                False   
1                False                     False                False   
2                False                    

### Payment Behavior

Payment Behavior is the last categorical feature that requires one hot encoding. Before we one hot encoded, we ensured that there were no NaN values, which were represented by !@9#%8, by replacing any of these values with the most occuring payment behavior type for a specific customer. After, we one hot encode this feature, which has several different options that relate to the behavior of a payment.

In [173]:
print(data['Payment_Behaviour'].value_counts())

grouped_data = data.groupby('Customer_ID')
data['Payment_Behaviour'] = grouped_data['Payment_Behaviour'].transform(lambda x: replace_invalid_with_mode(x, '!@9#%8'))

data.dropna(subset=['Payment_Behaviour'], inplace = True)

print(data['Payment_Behaviour'].value_counts())

# One hot encode:
occupation_dummies = pd.get_dummies(data['Payment_Behaviour'], prefix = 'Payment_Behaviour')
data = pd.concat([data, occupation_dummies], axis = 1)
data.drop('Payment_Behaviour', axis = 1, inplace = True)

data.head()

Payment_Behaviour
Low_spent_Small_value_payments      21085
High_spent_Medium_value_payments    14685
High_spent_Large_value_payments     11571
Low_spent_Medium_value_payments     11509
High_spent_Small_value_payments      9469
Low_spent_Large_value_payments       8662
!@9#%8                               6297
Name: count, dtype: int64
Payment_Behaviour
Low_spent_Small_value_payments      21847
High_spent_Medium_value_payments    15072
High_spent_Large_value_payments     11901
Low_spent_Medium_value_payments     11383
High_spent_Small_value_payments      9311
Low_spent_Large_value_payments       8498
Name: count, dtype: int64


Customer_ID   Age  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
0   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
1   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
2   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
3   CUS_0xd40  23.0       19114.12            1824.843333                3.0   
4   CUS_0xd40  23.0       19114.12            1824.843333                3.0   

   Num_Credit_Card  Interest_Rate  Num_of_Loan  Delay_from_due_date  \
0              4.0            3.0          4.0                  3.0   
1              4.0            3.0          4.0                  3.0   
2              4.0            3.0          4.0                  5.0   
3              4.0            3.0          4.0                  6.0   
4              4.0            3.0          4.0                  8.0   

   Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
0                     7.0                 11.27                   4.0   
1                     7.0                 11.27                   4.0   
2                     4.0                  6.27                   4.0   
3                     4.0                 11.27                   4.0   
4                     4.0                  9.27                   4.0   

   Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
0            809.98                 26.822620               265.0   
1            809.98                 28.609352               267.0   
2            809.98                 31.377862               268.0   
3            809.98                 24.797347               269.0   
4            809.98                 27.262259               270.0   

   Total_EMI_per_month  Amount_invested_monthly  Monthly_Balance Credit_Score  \
0            49.574949                80.415295       312.494089         Good   
1            49.574949                81.699521       331.209863         Good   
2            49.574949               199.458074       223.451310         Good   
3            49.574949                41.420153       341.489231         Good   
4            49.574949                62.430172       340.479212         Good   

   Occupation_Accountant  Occupation_Architect  Occupation_Developer  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   Occupation_Doctor  Occupation_Engineer  Occupation_Entrepreneur  \
0              False                False                    False   
1              False                False                    False   
2              False                False                    False   
3              False                False                    False   
4              False                False                    False   

   Occupation_Journalist  Occupation_Lawyer  Occupation_Manager  \
0                  False              False               False   
1                  False              False               False   
2                  False              False               False   
3                  False              False               False   
4                  False              False               False   

   Occupation_Mechanic  Occupation_Media_Manager  Occupation_Musician  \
0                False                     False                False   
1                False                     False                False   
2                False                     False                False   
3                False                     False                False   
4                False                     False                False   

   Occupation_Scientist  Occupation_Teacher  Occupation_U

Now that all the features have been thoroughly cleaned, we wanted to make sure that all the data types were good before continuing on. Next, we stored our labels in a variable, so that it is easily accessed later on as we will drop the column to preform steps for modeling. Customer_ID finally gets dropped as a feature because it is another feature that does not help tell us whether a customer has a good, bad, or standard credit score. Whenever, we one hot encode, having all columns can redundant because if we drop one of the columns for a one hot encoding of a feature, we know that the record belongs to the column if all other values for the feature is false. Once we check this, we are now ready to proceed to model and analyze machine learning results.

In [174]:
# Ensure all columns are proper types.
for col in data.columns:
    print(f'{col} : {pd.api.types.infer_dtype(data[col])}')
    
# Credit score column as labels column.
labels = data['Credit_Score'].copy()
data.drop(columns=['Credit_Score'], inplace = True)

# Drop Customer_ID because it is useless.
data.drop(columns=['Customer_ID'], inplace = True)

# TODO: Drop one-hot encoding columns that are "yes"/"no" since we only need 1.
data.drop(columns=['Occupation_Unemployed'], inplace = True)
data.drop(columns=['Not Specified'], inplace = True)
data.drop(columns=['Credit Mix_Standard'], inplace = True)
data.drop(columns=['Payment_of_Min_Amount_No'], inplace = True)
data.drop(columns=['Payment_Behaviour_Low_spent_Small_value_payments'], inplace = True)

print(data.shape[0])
data.head()

# TODO: PCA Analysis
# from sklearn.preprocessing import StandardScaler
# from sklearn.decomposition import PCA
# scaler = StandardScaler()
# scaler.fit(data)
# std_data = scaler.transform(data)
# pca = PCA(n_components = 0.95, svd_solver = 'full')
# pca_data = pca.fit_transform(std_data)
# pca_df = pd.DataFrame(pca_data)
# pca_df.head()

Customer_ID : string
Age : floating
Annual_Income : floating
Monthly_Inhand_Salary : floating
Num_Bank_Accounts : floating
Num_Credit_Card : floating
Interest_Rate : floating
Num_of_Loan : floating
Delay_from_due_date : floating
Num_of_Delayed_Payment : floating
Changed_Credit_Limit : floating
Num_Credit_Inquiries : floating
Outstanding_Debt : floating
Credit_Utilization_Ratio : floating
Credit_History_Age : floating
Total_EMI_per_month : floating
Amount_invested_monthly : floating
Monthly_Balance : floating
Credit_Score : string
Occupation_Accountant : boolean
Occupation_Architect : boolean
Occupation_Developer : boolean
Occupation_Doctor : boolean
Occupation_Engineer : boolean
Occupation_Entrepreneur : boolean
Occupation_Journalist : boolean
Occupation_Lawyer : boolean
Occupation_Manager : boolean
Occupation_Mechanic : boolean
Occupation_Media_Manager : boolean
Occupation_Musician : boolean
Occupation_Scientist : boolean
Occupation_Teacher : boolean
Occupation_Unemployed : boolean
Oc

Age  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
0  23.0       19114.12            1824.843333                3.0   
1  23.0       19114.12            1824.843333                3.0   
2  23.0       19114.12            1824.843333                3.0   
3  23.0       19114.12            1824.843333                3.0   
4  23.0       19114.12            1824.843333                3.0   

   Num_Credit_Card  Interest_Rate  Num_of_Loan  Delay_from_due_date  \
0              4.0            3.0          4.0                  3.0   
1              4.0            3.0          4.0                  3.0   
2              4.0            3.0          4.0                  5.0   
3              4.0            3.0          4.0                  6.0   
4              4.0            3.0          4.0                  8.0   

   Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
0                     7.0                 11.27                   4.0   
1                     7.0                 11.27                   4.0   
2                     4.0                  6.27                   4.0   
3                     4.0                 11.27                   4.0   
4                     4.0                  9.27                   4.0   

   Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
0            809.98                 26.822620               265.0   
1            809.98                 28.609352               267.0   
2            809.98                 31.377862               268.0   
3            809.98                 24.797347               269.0   
4            809.98                 27.262259               270.0   

   Total_EMI_per_month  Amount_invested_monthly  Monthly_Balance  \
0            49.574949                80.415295       312.494089   
1            49.574949                81.699521       331.209863   
2            49.574949               199.458074       223.451310   
3            49.574949                41.420153       341.489231   
4            49.574949                62.430172       340.479212   

   Occupation_Accountant  Occupation_Architect  Occupation_Developer  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   Occupation_Doctor  Occupation_Engineer  Occupation_Entrepreneur  \
0              False                False                    False   
1              False                False                    False   
2              False                False                    False   
3              False                False                    False   
4              False                False                    False   

   Occupation_Journalist  Occupation_Lawyer  Occupation_Manager  \
0                  False              False               False   
1                  False              False               False   
2                  False              False               False   
3                  False              False               False   
4                  False              False               False   

   Occupation_Mechanic  Occupation_Media_Manager  Occupation_Musician  \
0                False                     False                False   
1                False                     False                False   
2                False                     False                False   
3                False                     False                False   
4                False                     False                False   

   Occupation_Scientist  Occupation_Teacher  Occupation_Writer  \
0                  True               False              False   
1                  True               False              False   
2          

# Building the Model

In [175]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV

clf = DecisionTreeClassifier()
data_y = labels
data_x = data # Features for the data
params = {"max_depth": [10, 20, 30],
"min_samples_leaf": [5, 10, 15], "max_features": [5, 10, 15]}
grid_search = GridSearchCV(clf, params, cv=5, scoring='accuracy', n_jobs = -1)

grid_search.fit(data_x, data_y)

# What did it find as the best max_depth?
print("Best Parameters:", grid_search.best_params_)
# What was the accuracy at this best max_depth?
print(f'Accuracy: {grid_search.best_score_*100}%')

Best Parameters: {'max_depth': 10, 'max_features': 15, 'min_samples_leaf': 15}
Accuracy: 69.19835327602873%
